pallets/assets/src/lib.rs

```diff
@@ -0,0 +1,250 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+// #[cfg(test)]
+// mod mock;
+//
+// #[cfg(test)]
+// mod tests;
+//
+// #[cfg(feature = "runtime-benchmarks")]
+// mod benchmarking;
+pub mod weights;
+
+pub mod traits;
+
+use sp_runtime::traits::StaticLookup;
+
+pub use pallet::*;
+use sp_std::borrow::Borrow;
+use sp_std::prelude::*;
+pub use weights::WeightInfo;
+
+type AccountIdLookupOf<T> = <<T as frame_system::Config>::Lookup as StaticLookup>::Source;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+    use crate::traits::AssetValidator;
+    use frame_support::pallet_prelude::*;
+    use frame_system::pallet_prelude::*;
+    use xcm::latest::MultiLocation;
+    use xcm::prelude::{Abstract, AssetId, Concrete, GeneralIndex, PalletInstance, Parachain, X3};
+    use xcm_executor::traits::Convert;
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    #[pallet::without_storage_info]
+    pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
+
+    #[pallet::config]
+    pub trait Config<I: 'static = ()>: frame_system::Config + pallet_assets::Config<I> {
+        type RuntimeEvent: From<Event<Self, I>>
+            + IsType<<Self as pallet_assets::Config<I>>::RuntimeEvent>
+            + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+        /// Weight information for extrinsics in this pallet.
+        type WeightInfo: WeightInfo;
+    }
+
+    #[pallet::genesis_config]
+    pub struct GenesisConfig<T: Config<I>, I: 'static = ()> {
+        /// Genesis assets: `internal asset ID -> asset ID` (Statemint's general index)
+        // TODO generalize asset ID to any XCM AssetID once structs derive deserialize (merged with XCM-3)
+        pub assets: Vec<(<T as pallet_assets::Config<I>>::AssetId, u32)>,
+    }
+
+    #[cfg(feature = "std")]
+    impl<T: Config<I>, I: 'static> Default for GenesisConfig<T, I> {
+        fn default() -> Self {
+            Self { assets: vec![] }
+        }
+    }
+
+    #[pallet::genesis_build]
+    impl<T: Config<I>, I: 'static> GenesisBuild<T, I> for GenesisConfig<T, I> {
+        fn build(&self) {
+            for &(internal_asset_id, general_index) in &self.assets {
+                let asset_id = AssetId::Concrete(MultiLocation::new(
+                    1,
+                    X3(
+                        Parachain(1000),
+                        PalletInstance(50),
```

Comment:
@gitsimon we probably should not hardcode these values (1, 1000, 50), but get them from the `Config` instead.